packages/compiler/core/parser.ts

```diff
@@ -2421,3 +2423,42 @@ function isBlocklessNamespace(node: Node) {
 
   return node.statements === undefined;
 }
+
+export function getFirstAncestor(node: Node, test: NodeCallback<boolean>): Node | undefined {
+  for (let n = node.parent; n; n = n.parent) {
+    if (test(n)) {
+      return n;
+    }
+  }
+  return undefined;
+}
+
+export function getIdentifierContext(id: IdentifierNode): IdentifierContext {
+  const node = getFirstAncestor(id, (n) => n.kind !== SyntaxKind.MemberExpression);
+  compilerAssert(node, "Identifier with no non-member-expression ancestor.");
+
+  let kind: IdentifierKind;
+  switch (node.kind) {
+    case SyntaxKind.TypeReference:
+      kind = IdentifierKind.TypeReference;
+      break;
+    case SyntaxKind.DecoratorExpression:
+      kind = IdentifierKind.Decorator;
+      break;
+    case SyntaxKind.UsingStatement:
+      kind = IdentifierKind.Using;
+      break;
+    case SyntaxKind.DirectiveExpression:
+      kind = IdentifierKind.Directive;
+      break;
+    default:
+      compilerAssert(
+        (id.parent as any).id === id,
+        "This identifier does not appear to be a declaration. There is probably a missing switch case above."
+      );
+      kind = IdentifierKind.Declaration;
```

Comment:
Does this line need to be there, isn't the compileAssert going to crash the program?

Reply:
I would prefer to get this crash then to debug strange behavior that would result if we call this a declaration and it isn't.

Reply:
That said, I'm almost sure I've missed some things here, I'll see if I can run this over all the identifiers in our tests.

Reply:
oh... my bad, didn't see the compilerAssert had a condition and wasn't always on for this switch case

Reply:
Testing found lots of missed cases in projections. I'm changing this to set the kind to "Other" when this condition doesn't hold so we can add support over time.